wip/beam-spec.md

```diff
@@ -0,0 +1,437 @@
+# The Basic Embedded Abstract Machine (BEAM)
+
+## Summary
+
+This document contains the specification of an *abstract machine*. The goal of
+this specification is facilitating discussions about the soundness of `unsafe`
+concurrent code that runs on microcontrollers.
+
+The BEAM is meant to be representative of the microcontrollers one can run Rust
+code on, as such it has concurrency capabilities commonly found in this kind of
+hardware but exposes these capabilities in a way that's not tied to any
+particular architecture.
+
+## Program execution
+
+- The BEAM has a single core / processing-unit and can only run a single process
+  on this core.
+
+- When the BEAM executes a program it uses the function called `main` as the
+  entry point. Returning from this function is UNSPECIFIED behavior.
+
+- The signature of `main` must be `[unsafe] fn() [-> !]`. That is the program
+  takes no arguments.
+
+## Interrupt handling
+
+(Interrupt handling is a preemption mechanism that in its most basic form works
+as follows: when an interrupt *signal* arrives the processor suspends the
+execution of the current function and jumps to a special subroutine known as the
+interrupt *handler*. When this subroutine returns the processors resumes the
+execution of the function that was preempted by the interrupt handler.)
```

Comment:
I think this should not be just a parenthetical. This is important background for people not familiar with the area.

Reply:
This explains the "signal" and "interrupt handler" concepts. I think it might be worth it to also explain here "masking", since that term is used below a couple of times before being introduced.